File: 01220. Count Vowels Permutation/solution.go
/**
 * Problem: 1220. Count Vowels Permutation
 *
 * Difficulty: Hard
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func countVowelPermutation(n int) int {
	// Initialize counts for each vowel for strings of length 1
	var countA, countE, countI, countO, countU int64 = 1, 1, 1, 1, 1

	// Iterate for the remaining n-1 positions
	for n--; n > 0; n-- {
		// Update counts for each vowel based on the rules:
		// 'a' can follow 'e', 'i', 'u'
		// 'e' can follow 'a', 'i'
		// 'i' can follow 'e', 'o'
		// 'o' can follow 'i'
		// 'u' can follow 'i', 'o'
		countA, countE, countI, countO, countU = 
			(countE + countI + countU) % 1000000007, // new countA
			(countA + countI) % 1000000007,          // new countE
			(countE + countO) % 1000000007,          // new countI
			countI,                                  // new countO
			(countI + countO) % 1000000007           // new countU
	}

	// Return the total number of valid strings modulo 1e9+7
	return int((countA + countE + countI + countO + countU) % 1000000007)
}

File: 01220. Count Vowels Permutation/solution.js
/**
 * Problem: 1220. Count Vowels Permutation
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Returns the number of vowel permutations of length n
 *
 * @param {number} n - String length
 *
 * @returns {number} - Total permutations
 */
const countVowelPermutation = (n) => {
  // Define the modulo constant to prevent integer overflow
  const MOD = 1e9 + 7

  // Initialize the count for each vowel for strings of length 1
  let countA = 1, // Number of strings ending with 'a'
    countE = 1, // Number of strings ending with 'e'
    countI = 1, // Number of strings ending with 'i'
    countO = 1, // Number of strings ending with 'o'
    countU = 1 // Number of strings ending with 'u'

  // Iterate from the second character to the nth character
  for (let index = 1; index < n; index++) {
    // Calculate next counts based on vowel transition rules
    const nextA = countE // 'a' follows only 'e'
    const nextE = (countA + countI) % MOD // 'e' follows 'a' and 'i'
    const nextI = (countA + countE + countO + countU) % MOD // 'i' follows all except 'i'
    const nextO = (countI + countU) % MOD // 'o' follows 'i' and 'u'
    const nextU = countA // 'u' follows only 'a'

    // Update counts for the next iteration
    countA = nextA
    countE = nextE
    countI = nextI
    countO = nextO
    countU = nextU
  }

  // Return the total number of valid strings modulo MOD
  return (countA + countE + countI + countO + countU) % MOD
}
